crates/noirc_evaluator/src/ssa/conditional.rs

```diff
@@ -485,6 +481,153 @@ impl DecisionTree {
+            }
+            //merge i:
+            let ins = ctx.get_instruction(left[left_pos]);
+            let jins = ctx.get_instruction(right[right_pos]);
+            let assumption = &self[ctx[block_id].assumption];
+
+            let mut to_add = Vec::new();
+            let mut merged_op = match &ins.operation {
+                Operation::Call { arguments, .. } => {
+                    if let Operation::Call { func_id, arguments: a2, returned_arrays, .. } =
+                        &jins.operation
+                    {
+                        for a in arguments.iter().enumerate() {
+                            let op = Operation::Cond {
+                                condition: self[assumption.parent].condition,
+                                val_true: *a.1,
+                                val_false: a2[a.0],
+                            };
+                            let typ = ctx.get_object_type(*a.1);
+                            to_add.push(Instruction::new(op, typ, Some(block_id)));
+                        }
+                        Operation::Call {
+                            func_id: *func_id,
+                            arguments: Vec::new(),
+                            returned_arrays: returned_arrays.clone(),
+                            predicate: self.root,
+                        }
+                    } else {
+                        unreachable!();
+                    }
+                }
+                Operation::Store { array_id, index, value } => {
+                    if let Operation::Store { value: j_value, .. } = jins.operation {
+                        let op = Operation::Cond {
+                            condition: self[assumption.parent].condition,
+                            val_true: *value,
+                            val_false: j_value,
+                        };
+                        let merge =
+                            Instruction::new(op, ctx.mem[*array_id].element_type, Some(block_id));
+                        to_add.push(merge);
+                        // TODO: we should simplify this instruction
+                    }
+                    Operation::Store { array_id: *array_id, index: *index, value: NodeId::dummy() }
+                }
+                _ => unreachable!(),
+            };
+            let mut to_add_id = Vec::new();
+            for merge in to_add {
+                let merge_id = ctx.add_instruction(merge);
+                result.push(merge_id);
+                to_add_id.push(merge_id);
+            }
+            if let Operation::Store { value, .. } = &mut merged_op {
+                *value = *to_add_id.last().unwrap();
+            } else {
+                if let Operation::Call { arguments, .. } = &mut merged_op {
+                    *arguments = to_add_id;
+                }
+                let jins1 = ctx.get_mut_instruction(left[left_pos]);
+                jins1.mark = node::Mark::ReplaceWith(right[right_pos]);
```

Comment:
Why replace left with right here?

Reply:
Because we keep the right, but if some code was using the left, it must references the instruction we keep now.
Of course we could have done the opposite; keeping the left and removing the right.
I do not do it for Store because they do not have a result so nobody should be using them.